lib/srv/sess.go

```diff
@@ -810,8 +815,10 @@ func (s *session) emitSessionLeaveEvent(ctx *ServerContext) {
 }
 
 // emitSessionEndEvent emits a session end event.
-// Must be called under session Lock.
 func (s *session) emitSessionEndEvent() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
Turns out this alone would have fixed the data race... But the other changes still prevent other potential race conditions.